blog/2022/04/simple-chains.md

```diff
@@ -0,0 +1,489 @@
+ 70.517619 seconds (98.27 M allocations: 67.833 GiB, 3.94% gc time, 39.30% compilation time)
+
+julia> eval_loss_accuracy(train_loader, model, device),
+       eval_loss_accuracy(test_loader, model, device)
+((loss = 0.1306f0, acc = 96.14), (loss = 0.1144f0, acc = 96.57))
+
+julia> @time train!(model, train_loader)
+ 44.241192 seconds (369.06 k allocations: 62.864 GiB, 2.55% gc time)
+
+julia> eval_loss_accuracy(train_loader, model, device),
+       eval_loss_accuracy(test_loader, model, device)
+((loss = 0.072f0, acc = 97.7883), (loss = 0.0642f0, acc = 97.99))
+```
+
+#### How long did SimpleChains.jl take?
+
+SimpleChains on an AMD EPYC 7513 32-Core Processor:
+
+```julia
+julia> @time SimpleChains.train_batched!(G, p, lenetloss, xtrain4, SimpleChains.ADAM(3e-4), 10);
+ 34.410432 seconds (55.84 M allocations: 5.920 GiB, 3.79% gc time, 85.95% compilation time)
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtrain4, p)
+(0.972, 0.093898475f0)
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtest4, ytest1, p)
+(0.9744, 0.08624289f0)
+
+julia> @time SimpleChains.train_batched!(G, p, lenetloss, xtrain4, SimpleChains.ADAM(3e-4), 10);
+  3.083624 seconds
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtrain4, p)
+(0.9835666666666667, 0.056287352f0)
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtest4, ytest1, p)
+(0.9831, 0.053463124f0)
+```
+
+SimpleChains on an Intel i9 10980XE 18-Core Processor:
+
+```julia
+julia> @time SimpleChains.train_batched!(G, p, lenetloss, xtrain4, SimpleChains.ADAM(3e-4), 10);
+ 34.915917 seconds (83.99 M allocations: 5.431 GiB, 3.91% gc time, 95.22% compilation time)
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtrain4, p)
+(0.972, 0.09354472f0)
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtest4, ytest1, p)
+(0.975, 0.084895514f0)
+
+julia> @time SimpleChains.train_batched!(G, p, lenetloss, xtrain4, SimpleChains.ADAM(3e-4), 10);
+  1.324148 seconds
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtrain4, p)
+(0.9836833333333334, 0.054299146f0)
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtest4, ytest1, p)
+(0.9829, 0.053510584f0)
+```
+
+SimpleChains on an Intel i7 1165G7 4-Core Processor (thin and light laptop CPU):
+
+```julia
+julia> @time SimpleChains.train_batched!(G, p, lenetloss, xtrain4, SimpleChains.ADAM(3e-4), 10);
+ 42.664702 seconds (103.71 M allocations: 5.236 GiB, 2.85% gc time, 77.60% compilation time)
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtrain4, p)
+(0.9863166666666666, 0.044742946f0)
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtest4, ytest1, p)
+(0.9841, 0.048268355f0)
+
+julia> @time SimpleChains.train_batched!(G, p, lenetloss, xtrain4, SimpleChains.ADAM(3e-4), 10);
+  6.320631 seconds
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtrain4, p)
+(0.99485, 0.017716348f0)
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtest4, ytest1, p)
+(0.9888, 0.035394628f0)
```

Comment:
```suggestion
julia> @time SimpleChains.train_batched!(G, p, lenetloss, xtrain, SimpleChains.ADAM(3e-4), 10);
 41.053800 seconds (104.10 M allocations: 5.263 GiB, 2.83% gc time, 77.62% compilation time)

julia> SimpleChains.accuracy_and_loss(lenetloss, xtrain, p),
       SimpleChains.accuracy_and_loss(lenetloss, xtest, ytest, p)
((0.9491333333333334, 0.16993132f0), (0.9508, 0.15890576f0))

julia> @time SimpleChains.train_batched!(G, p, lenetloss, xtrain, SimpleChains.ADAM(3e-4), 10);
  5.320512 seconds

julia> SimpleChains.accuracy_and_loss(lenetloss, xtrain, p),
       SimpleChains.accuracy_and_loss(lenetloss, xtest, ytest, p)
((0.9700833333333333, 0.10100537f0), (0.9689, 0.09761506f0))
```

Reply:
My poor laptop has a hard time taking the heat, runs out of power budget, or something that causes it to begin throttling.
So I didn't copy/paste the script as one giant block, but waited a few seconds before running the second `@time SimpleChains.train_batched!` to let it cool off. Might be worth mentioning this.